File: LeetCode/remote/combination_sum_III.cs
using System;
using System.Collections.Generic;
using System.Linq;

// You are here!
// Your runtime beats -0.00% of csharp coders. <- Aahahahahahaha!!!
//
// https://leetcode.com/submissions/detail/38153904/
//
// 
// Submission Details
// 18 / 18 test cases passed.
//  Status: Accepted
//  Runtime: 460 ms
//      
//      Submitted: 0 minutes ago
//

static class Program {
    public static IList<IList<int>> CombinationSum3(int k, int n) {
         IList<IList<int>> result = new List<IList<int>>();
         Combinations(0, k, 1, n, result, new List<int>());
         return result;
    }

    private static void Combinations(int sum, int k, int count, int n, IList<IList<int>> result, List<int> partial) {
        if (sum == n && partial.Count == k) {
            result.Add(new List<int>(partial));
            return;
        }

        for (int i = count; i <= 9; i++) {
            if (sum + i <= n && partial.Count < k) {
                partial.Add(i);
                Combinations(sum + i, k, i + 1, n, result, partial);
                partial.Remove(i);
            }
        }
    }

    private static String GetSome(this IList<IList<int>> list) {
        return String.Join("; ", list.Select(x => String.Join(", ", x)));
    }

    public static void Main() {
        Console.WriteLine(CombinationSum3(3, 9).GetSome());
        Console.WriteLine(CombinationSum3(0, 9).GetSome());
        Console.WriteLine(CombinationSum3(0, 0).GetSome());
        Console.WriteLine(CombinationSum3(1, 1).GetSome());
        Console.WriteLine(CombinationSum3(3, 15).GetSome());
    }
}

File: LeetCode/remote/combination_sum_III.rb
#   https://leetcode.com/problems/combination-sum-ii/
#
#    Given a collection of candidate numbers (C) and a target number (T), find all unique combinations in C where the candidate numbers sums to T.
#
#    Each number in C may only be used once in the combination.
#    
#    https://leetcode.com/submissions/detail/54457255/
#   
#   Submission Details
#   172 / 172 test cases passed.
#       Status: Accepted
#       Runtime: 172 ms
#           
#           Submitted: 0 minutes ago

def combination_sum2(candidates, target)
    combination_sum(candidates.sort, target)
end

def combination_sum(candidates, target)
    return nil if target < 0
    result = []
    (0...candidates.size).each { |i|
        next if i > 0 && candidates[i] == candidates[i - 1]
        candidate = candidates[i]
        if candidate == target
            result << [candidate]
        else
            right = combination_sum(candidates[i + 1..-1], target - candidate)
            next if right.nil?
            
            right.each { |x|
                result << [candidate] + x
            }
        end
    }
    
    result
end
